File: booksamples/chapter-13/src/main/java/org/vulkanb/eng/graph/vk/FrameBuffer.java
package org.vulkanb.eng.graph.vk;

import org.lwjgl.system.MemoryStack;
import org.lwjgl.vulkan.VkFramebufferCreateInfo;

import java.nio.LongBuffer;

import static org.lwjgl.vulkan.VK11.*;
import static org.vulkanb.eng.graph.vk.VulkanUtils.vkCheck;

public class FrameBuffer {

    private final Device device;
    private final long vkFrameBuffer;

    public FrameBuffer(Device device, int width, int height, LongBuffer pAttachments, long renderPass, int layers) {
        this.device = device;
        try (MemoryStack stack = MemoryStack.stackPush()) {
            VkFramebufferCreateInfo fci = VkFramebufferCreateInfo.calloc(stack)
                    .sType(VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO)
                    .pAttachments(pAttachments)
                    .width(width)
                    .height(height)
                    .layers(layers)
                    .renderPass(renderPass);

            LongBuffer lp = stack.mallocLong(1);
            vkCheck(vkCreateFramebuffer(device.getVkDevice(), fci, null, lp),
                    "Failed to create FrameBuffer");
            vkFrameBuffer = lp.get(0);
        }
    }

    public void cleanup() {
        vkDestroyFramebuffer(device.getVkDevice(), vkFrameBuffer, null);
    }

    public long getVkFrameBuffer() {
        return vkFrameBuffer;
    }

}

File: booksamples/chapter-05/kotlin/src/main/kotlin/org/vulkanb/eng/graph/vk/FrameBuffer.kt
package org.vulkanb.eng.graph.vk

import org.lwjgl.system.MemoryStack
import org.lwjgl.vulkan.VK11.*
import org.lwjgl.vulkan.VkFramebufferCreateInfo
import org.vulkanb.eng.graph.vk.VulkanUtils.vkCheck
import java.nio.LongBuffer

class FrameBuffer(private val device: Device, width: Int, height: Int, pAttachments: LongBuffer?, renderPass: Long) {
    var vkFrameBuffer: Long = 0

    init {
        MemoryStack.stackPush().use { stack ->
            val fci = VkFramebufferCreateInfo.calloc(stack)
                .sType(VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO)
                .pAttachments(pAttachments)
                .width(width)
                .height(height)
                .layers(1)
                .renderPass(renderPass)
            val lp = stack.mallocLong(1)
            vkCreateFramebuffer(device.vkDevice, fci, null, lp)
                .vkCheck("Failed to create FrameBuffer")
            vkFrameBuffer = lp[0]
        }
    }

    fun cleanup() {
        vkDestroyFramebuffer(device.vkDevice, vkFrameBuffer, null)
    }
}